### ISSUE TITLE
Temporary arrays not allocated

### ISSUE THREAD

**Issue by EmilyBourne** [pyccel/pyccel#1130]
**Describe the bug**
When passing an array created on the fly, to a function, the array creation is not allocated or unravelled into loops. This is fine in languages which handle vector operations, but not in languages such as C.

**To Reproduce**
Provide code to reproduce the behaviour:
```python
def print_a(a : 'int[:]'):
    print(a)

def print_sum(a : 'int[:]', b : 'int[:]'):
    print_a(a+b)
```

Provide the generated code, or the error message:
```c
/*........................................*/
void print_a(t_ndarray a)
{
    int64_t i;
    printf("%s", "[");
    for (i = 0; i < a.shape[0] - 1; i += 1)
    {
        printf("%ld ", GET_ELEMENT(a, nd_int64, (int64_t)i));
    }
    printf("%ld]\n", GET_ELEMENT(a, nd_int64, (int64_t)a.shape[0] - 1));
}
/*........................................*/
/*........................................*/
void print_sum(t_ndarray a, t_ndarray b)
{
    print_a(a + b);
}
/*........................................*/
```

**Expected behaviour**
The creation of the temporary should be moved out of the function call at the semantic stage so that the addition is correctly unrolled into loops.
The code should be translated to:
```c
/*........................................*/
void print_a(t_ndarray a)
{
    int64_t i;
    printf("%s", "[");
    for (i = 0; i < a.shape[0] - 1; i += 1)
    {
        printf("%ld ", GET_ELEMENT(a, nd_int64, (int64_t)i));
    }
    printf("%ld]\n", GET_ELEMENT(a, nd_int64, (int64_t)a.shape[0] - 1));
}
/*........................................*/
/*........................................*/
void print_sum(t_ndarray a, t_ndarray b)
{
    t_ndarray Dummy_0001 = {.shape = NULL};
    int64_t i;
    Dummy_0001 = array_create(1, (int64_t[]){a.shape[0]}, nd_int64, false);
    for (i = 0; i < a.shape[0]; i += 1)
    {
        GET_ELEMENT(Dummy_0001, nd_int64, (int64_t)i) = GET_ELEMENT(a, nd_int64, (int64_t)i) + GET_ELEMENT(b, nd_int64, (int64_t)i);
    }
    print_a(Dummy_0001);
    free_array(Dummy_0001);
}
/*........................................*/
```

**Language**
C

**Comment by EmilyBourne** [comment]
This issue is almost identical to issue #1126 . They should be fixed together

**Comment by Pinkyboi** [comment]
@EmilyBourne I think this issue was solved by [#1187](https://github.com/pyccel/pyccel/pull/1187).

**Comment by EmilyBourne** [comment]
I agree except in the case described by #1126
